// @dagger.io/dagger/dist/common/errors/errors-codes.js
export const ERROR_CODES = {
    /**
     * {@link GraphQLRequestError}
     */
    GraphQLRequestError: "D100",
    /**
     * {@link UnknownDaggerError}
     */
    UnknownDaggerError: "D101",
    /**
     * {@link TooManyNestedObjectsError}
     */
    TooManyNestedObjectsError: "D102",
    /**
     * {@link EngineSessionConnectParamsParseError}
     */
    EngineSessionConnectParamsParseError: "D103",
    /**
     * {@link EngineSessionConnectionTimeoutError}
     */
    EngineSessionConnectionTimeoutError: "D104",
    /**
     * {@link EngineSessionError}
     */
    EngineSessionError: "D105",
    /**
     * {@link InitEngineSessionBinaryError}
     */
    InitEngineSessionBinaryError: "D106",
    /**
     * {@link DockerImageRefValidationError}
     */
    DockerImageRefValidationError: "D107",
    /**
     * {@link NotAwaitedRequestError}
     */
    NotAwaitedRequestError: "D108",
    /**
     * (@link ExecError}
     */
    ExecError: "D109",
};
export const ERROR_NAMES = Object.keys(ERROR_CODES).reduce((obj, item) => (Object.assign(Object.assign({}, obj), { [item]: item })), {});
