// @lifeomic/axios-fetch/src/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildAxiosFetch = void 0;
const typeUtils_1 = require("./typeUtils");
/**
 * A Fetch WebAPI implementation based on the Axios client
 */
const axiosFetch = (axios, 
// Convert the `fetch` style arguments into a Axios style config
transformer = (config) => config) => async (input, init) => {
    const rawHeaders = (0, typeUtils_1.createAxiosHeaders)(init === null || init === void 0 ? void 0 : init.headers);
    const lowerCasedHeaders = {};
    Object.entries(rawHeaders).forEach(([name, value]) => {
        lowerCasedHeaders[name.toLowerCase()] = value;
    });
    if (!('content-type' in lowerCasedHeaders)) {
        lowerCasedHeaders['content-type'] = 'text/plain;charset=UTF-8';
    }
    const rawConfig = {
        url: (0, typeUtils_1.getUrl)(input),
        method: (init === null || init === void 0 ? void 0 : init.method) || 'GET',
        data: init === null || init === void 0 ? void 0 : init.body,
        headers: lowerCasedHeaders,
        // Force the response to an arraybuffer type. Without this, the Response
        // object will try to guess the content type and add headers that weren't in
        // the response.
        // NOTE: Don't use 'stream' because it's not supported in the browser
        responseType: 'arraybuffer'
    };
    const config = transformer(rawConfig, input, init);
    let result;
    try {
        result = await axios.request(config);
    }
    catch (err) {
        if (err.response) {
            result = err.response;
        }
        else {
            throw err;
        }
    }
    return new Response(result.data, {
        status: result.status,
        statusText: result.statusText,
        headers: (0, typeUtils_1.createFetchHeaders)(result.headers)
    });
};
const buildAxiosFetch = (axios, transformer) => axiosFetch(axios, transformer);
exports.buildAxiosFetch = buildAxiosFetch;
